RollingDice finished

MilanG-git opened the pull request:
Hi all,
This is my solution for RollingDice. Here I try out for the first time the class method and play a bit around with verbose output.
Cheers,
Milan

MilanG-git commented:
Added my solution for MonoShuffle. Additionally, it shows the shuffle process in verbose mode.

Lorenzo360 commented:
Hi Milan, 
nice that you started using classes. That helped with reading the code. I have some feedback for you 

A4/MilanG-git-MonoShuffle.py 
  Line38:  self.seq = self.seq[:i] + nuc_j + self.seq[i+1:j] + nuc_i + self.seq[j+1:]
    Make more simple and efficient .... self.seq[j]=nuc_i and self.seq[i]=nuc_j

A4/MilanG-git-RollingDice.py
   Nice solution. I should have thought about using random.choices as well. Then it is really just one line ;) 



A4/MilanG-git-KLetShuffle.py
  Line 68:   def tree(self, start, u, path): why are you defining a function in a function?
Some hints for solving the shuffling problem (I took me a while to figure it out): 
--> Check the in- and out-degree of each node (only if the in- and out-degree of a node differ the shuffled sequence must start with the same letter(s) as the input sequence) 
--> Constructing the spanning tree can be tricky (in my case I had to debug endlessly)
--> I used a quit different approach with hash maps (dictionaries in python) and there it helped me to introduce artificial edges and nodes in case the in- and out-degrees differ for two nodes. Maybe this helps you too. 
E.g.: with -k=2 ---> XGCCACCAGAGAUGAGCGAAAGCCAACGCUUCUAUGAGCAGCGGCCGACGACUUGACUUACUUGGUGACGUGCCCGGCCCCCUGUAACCGCCCUGUACAACX
with -k=3 --> XXGCCACCAGAGAUGAGCGAAAGCCAACGCUUCUAUGAGCAGCGGCCGACGACUUGACUUACUUGGUGACGUGCCCGGCCCCCUGUAACCGCCCUGUACAACXX
etc. 
This guarantees that the in and out degree of each node is equal and made it much easier to create the spanning tree. 

I hope this helps, 
Good luck, 
Lorenz

MilanG-git commented:
Hi Lorenz,

Thanks for your feedback.

@ MonoShuffle: I'm using a string instead of a list for the sequence. I'm wondering if it would be more efficient to convert the sequence first into a list, to use the more simple way and then convert it back into a string?

@ RollingDice: Thanks.

@ KLetShuffle: In Line 68 I have the function inside another function to have the whole thing more compact. It made sense for me to have it in one place instead of two separate when writing the code.
I'm beginning to think that a dictionary would be the better data structure to find paths through the k-lets. Although, I believe that I'm not utilising the structure from the slides fully and that there is a trick missing, that I'm not yet seeing.
Thanks for your hints, they will definitely help as I try to figure out to properly construct the spanning tree :).

Cheers,
Milan

MilanG-git commented:
I implemented Wilson's algorithm to generate a spanning tree, as I understand it now better, as well as a function to check whether all nodes have been seen.

Thanks @robnic2718 , I found the error in sequence generation and fixed it.
